### PR TITLE
[Feature] pdf preview #2993

### PR DESCRIPTION
**Pull request by bkcsoft**
Implements PDF Previewer as per #2993 

Obligatory Screenshot:
2016-04-20-132321_1596x861_scrot

### CODE REVIEW

**Comment by tforgione** [comment]
Yep, this works correctly for me !

In fact, it's even better than github's method since they render the pdf to a canvas, and therefore, no search or copying from pdf is available !

**Comment by bkcsoft** [comment]
@Unknwon Thoughts?

**Comment by Unknwon** [comment]
Looks awesome, but 371 file changes?

**Comment by lunny** [comment]
Maybe pdf.js could be a git submodule on gogs so that to keep it update?

**Comment by lenovouser** [comment]
Maybe even consider using NPM for managing your JavaScript dependencies instead of submodules.

**Comment by andreynering** [comment]
My opinion is we should let the browser take care of it.

Firefox actually use pdf.js to handle any PDF file. Chrome has it's own viewer, too. Others falls back to what editor is available in the OS (e.g. Adobe Reader, Foxit, whatever).

Anyway it's better than including 371 files (mostly binary), since the browser already handles it.

So, I think just including an iframe with the PDF URL is enough.

**Comment by bkcsoft** [comment]
@Unknwon loads of locale and cmaps-files :unamused: 

@lunny hate submodules but it is indeed an idea
@lenovouser nope, not even once :rage:

@andreynering the browser can't take care of it if we want it inside the file-list (see the corresponding Issue). if is already using an iframe though (easier and better IMO (since contained JS-execution)), and you just reminded me that this always runs in Offline Mode, even if unchecked (should use https://mozilla.github.io/pdf.js/web/viewer.html when not in Offline Mode...)

**Comment by bkcsoft** [comment]
@lenovouser and in any case that would be bower, not npm, since npm is for node-modules while bower is for browser-modules

**Comment by Unknwon** [comment]
This is ... hard to merge.

**Comment by bkcsoft** [comment]
@Unknwon I could remove the locales and cmaps (risking breaking PDFs...) and we're down to 4-5 files...

**Comment by Unknwon** [comment]
What do `cmaps` do?

**Comment by bkcsoft** [comment]
@Unknwon https://blog.idrsolutions.com/2012/05/understanding-the-pdf-file-format-embedded-cmap-tables/

**Comment by Unknwon** [comment]
How many files left if remove locales?

**Comment by bkcsoft** [comment]
there are 169 cmaps, 105 locales, 6 js-files, 1 css, 1 html and 84 images (icons for buttons)

**Comment by bkcsoft** [comment]
the entire file-tree: https://github.com/bkcsoft/gogs/tree/feature/pdf-preview/public/plugins/pdfjs-1.4.20

**Comment by Unknwon** [comment]
I think remove locales could be the first step.

**Comment by bkcsoft** [comment]
@Unknwon done

**Comment by Unknwon** [comment]
>https://blog.idrsolutions.com/2012/05/understanding-the-pdf-file-format-embedded-cmap-tables/

From what I understand from this article, the CMAP tables are used for copy/paste/search? But display would be correct?

**Comment by bkcsoft** [comment]
It's also possible to do it like Github... with something like this... http://mozilla.github.io/pdf.js/examples/learning/prevnext.html

**Comment by bkcsoft** [comment]
The problem with doing it like Github is keeping track of multiple canvas-elements :trollface:

**Comment by Unknwon** [comment]
How bad it is without including CMAP tables/

**Comment by bkcsoft** [comment]
Seems to be working... 

2016-04-27-024308_1596x861_scrot

File: routers/repo/view.go [hunk 1]
```diff
@@ -92,9 +91,12 @@ func Home(ctx *context.Context) {
 
 			_, isTextFile := base.IsTextFile(buf)
 			_, isImageFile := base.IsImageFile(buf)
+			_, isPdfFile := base.IsPdfFile(buf)
```
Comment (Unknwon) [inline routers/repo/view.go]: `IsPDFFile` feels better? Same as below.

**Comment by bkcsoft** [comment]
@Unknwon Done

File: modules/base/tool.go [hunk 2]
```diff
@@ -522,3 +522,11 @@ func IsImageFile(data []byte) (string, bool) {
 	}
 	return contentType, false
 }
+
+func IsPdfFile(data []byte) (string, bool) {
```
Comment (Unknwon) [inline modules/base/tool.go]: Hmm?

**Comment by Unknwon** [comment]
Also, clean up to 1 commit would be the best.

**Comment by bkcsoft** [comment]
Done. And I was just keeping the history while fiddling around :trollface:

**Comment by Unknwon** [comment]
Thanks!

**Pull request merged**